Fix assumption about underflow on `AbstractUnitrange{<:Unsigned}`

Seelengrab opened the pull request:
As mentioned in #42528 after the initial PR, just removing `ifelse` is not enough. This fixes the behavior for creating empty UnitRanges of unsigned integers that don't underflow.

Seelengrab committed:
Add correct length/stop calculation for unsigned UnitRanges that don't assume overflow

As mentioned in #42528 after the initial PR, just removing `ifelse` is not enough. This fixes the behavior for creating empty UnitRanges of unsigned integers that don't overflow.

Seelengrab committed:
Unify `unitrange_last` for all `<:Integer` (#42528)

Seelengrab commented:
Failures seem unrelated (Task errors due to EOF?)

vtjnash reviewed (commented):
argh, I wrote a review thought, and forgot to finish it

Seelengrab committed:
Incorporate PR feedback

The new version enables branch elimination for `<: Signed`,
only introducing a branch for `<: Unsigned`, which need it.
